tensorflow/core/kernels/BUILD

```diff
@@ -52,7 +52,8 @@ load(
     "//third_party/mkl:build_defs.bzl",
     "if_mkl",
 )
-load("@local_config_cuda//cuda:build_defs.bzl", "if_cuda")
+load("@local_config_cuda//cuda:build_defs.bzl", "if_cuda", "if_cuda_is_configured")
+load("@local_config_rocm//rocm:build_defs.bzl", "if_rocm", "if_rocm_is_configured")
```

Comment:
> Are we planning to create the internal equivalents of these macros, just like what we have for CUDA?

It's @Artem-B and my opinion that `--config=cuda` (and by corollary `if_cuda`) is a harmful hack.  TF uses it as a switch to say "build for GPU or not", but it shouldn't: This should be decided by BUILD rules.  (E.g. you depend on `:tensorflow_cpu` or `:tensorflow_gpu`.)

The reason we introduced `--config=cuda` is because at the time, Skylark was missing some features we needed in order to make the toolchain work properly.  We think these features are there now.

Unfortunately the direction of these patches -- including the Eigen patch -- sort of doubles down on the notion of --config=cuda.  If we can't build a TF which includes both cuda and rocm bits, then we'll effectively never be able to get rid of --config=cuda.

That's why in XLA we've insisted that we maintain the ability to build for both cuda and rocm, determined by BUILD dependencies.

That said, it's unclear to me what is the alternative to these macros for TF, given that the structure of the rocm Eigen patches does not (?) allow us to build for AMDGPU and NVGPU in the same binary.  So TF may need its own versions of them internally; I don't see how else to do it.

Reply:
@jlebar Unfortunately given how GPU common runtime is designed I think it's hard to let TF be configured and built with both CUDA and ROCm at the same time. In `gpu_device.cc`, `EigenCudaStreamDevice` has a compile-time dependency to CUDA constructs. In my current implementation for ROCm, I renamed `EigenCudaStreamDevice` to `EigenGpuStreamDevice` and use `TENSORFLOW_USE_ROCM` macro to switch to ROCm-functional equivalents.

For XLA compiler, it's relatively easy to specify a new set of compiler backend and let it target AMDGPU. But for GPU common runtime, a bigger overhaul might be required if the ultimate goal is to get rid of `--config=cuda`.

That said, I believe such effort to modularize TF runtime, should be deferred to future PRs, after we have better consensus on how that be achieved?

Reply:
Let me try to clarify the way I'd like to see things working.
The goal is to be able to build any (and all) variants without specifying any extra bazel flags.
I.e. one should be able to say `bazel  //my_project:app_cuda //my_project:app_rocm //my_project:app_cpu` and get all three executables.

Let's suppose the app consists of main.cpp, kernel.cu (for CUDA and ROCm) and kernel_cpu.cpp (for CPU-only). All three app variants would use the same main.o, app_cuda will use kernel-cuda.o built from kernel.cu using CUDA-specific options/defines, kernel-rocm.o from kernel.cu using ROCm-specific options/defines and kernel-cpu would be compiled form kernel_cpu.cpp. User should be able to build any combination of them simultaneously. When you change anything in the build system, there's only one build configuration to test. If you've built an app there's no confusion about what it supports (or does not). I can't count number of times when someone attemted to run CPU-only TF and complained that it does not sees the GPUS or ran CUDA-enabled GPU on machine without GPUs and complained that it failed. Single build configuration  also saves on overall build time as the objects that don't care about CUDA/ROCm will be built only once, instead of once per build config. Considering that CUDA files constitute relatively small subset of tensorflow,  the difference is substantial. 

--config=cuda was inherited from internal Google build and exists for number of reasons that are not relevant to open-source tensorflow. I understand that it is convenient to continue adding extra dimensions to config parameters, but now that we're adding support for another accelerator is the good time to make sure we do it right. Maintaining and debugging multiple build configurations is a royal pain. Having single set of build rules for everything makes things somewhat easier to deal with.

Reply:
@Artem-B / @jlebar let me see if i understand this correctly to assess the efforts. It seems you expect one `bazel build` to get executables for all specified targets, and ditch `--config=XXX`.

As a corollary, would the following command be good for you? 1 bazel build get 3 PIP packages:

```
# 1 bazel build get 3 PIP packages
bazel build //tensorflow/tools/pip_package:build_cpu_pip_package //tensorflow/tools/pip_package:build_cuda_pip_package //tensorflow/tools/pip_package:build_rocm_pip_package
```

Also I'm wondering how shall we deal with test targets in `bazel test`?

Reply:
That's the idea. User tells **what** needs to be built, bazel and whoever maintains the build files figure out **how**.
As for the tests, it should follow roughly the same path.
We currently have rules that automatically generate CPU and GPU test variants already (i.e we generate test_foo for CPU and test_foo_gpu for CUDA), so, in a way, tests should mostly do the right thing already. Running correct set of tests will be similar to what happens now, modulo new suffixes on test targets.

Reply:
@gregstren sorry for inviting you to an new discussion thread without prior notification.

I, @whchung, am discussing with @jlebar / @Artem-B / @gunan about how to improve the build system of TensorFlow to make it to support multiple CPU / GPU platforms. And I've discovered your work of `dynamic bazel configurations` [1], `skylark build configurations` [2], and `2018 Bazel Configurability / Multiplatform Roadmap` [3] which eventually led me to bazel roadmaps pages [4] [5].

- [1] https://docs.google.com/document/d/1uoU8t7loTOu6uyzez-ilhcYgEXg4xjViJu0aqrf69TY/edit#
- [2] https://docs.google.com/document/d/1vc8v-kXjvgZOdQdnxPTaV0rrLxtP2XwnD2tAZlYJOqw/edit#
- [3] https://docs.google.com/document/d/1p7pdEwqnLOC-ATl-FdyHBJdk4Ae-5GzAYDvpJeMh4tU/edit?ts=5ac517b9#
- [4] https://www.bazel.build/roadmaps/configuration.html
- [5] https://www.bazel.build/roadmaps/platforms.html

I'm very new to the implementation of `bazel`, let alone how to test / adapt all these upcoming features of it in TensorFlow. I'm wondering could you help share some working example projects with targets tied to different toolchains, so I can learn better from those examples? Thank you very much.

Reply:
Sorry, I've completely forgot that OSS TF still uses custom crosstool to compile with nvcc. :-(
That's going to be a problem. Crosstool is not something that can be switched per build target,
and NVCC does not support clang as the host compiler. That forces us to have multiple build configurations. We still would be able to do something like `bazel --config=nvcc foo_cuda foo_cpu`, or `bazel foo_rocm foo_cpu` (probably with something like `--config=rocm`) but we will not be able to do all three at once. This complicates things.

I'll need to think about it a bit.

@gregestren -- It sounds like bazel has grown a lot of features lately that I've been missing for CUDA compilation. I'm glad to see that my handwavy proposal for TF compilation seems to be roughly in line with the general direction bazel configurability roadmap seems to head towards. I'll need to take a closer look at the recent changes to get better idea what we can do these days, but it appears that I may have more tools at my disposal than I used to in the past.

Reply:
Ok to merge then?

Reply:
Considering that at the moment we do not have a working alternative, that's probably the least bad option.
It does entangle the build with multiple crosstools, but it's a marginal bump in amount of work we'll need to do in addition to what's needed to deal with --config=nvcc. Making sure rocm build remains working will be a bit of a pain, but I expect bulk of the issues to be figured out on nvcc build, so overall it should not be a major issue.

I'm OK with the patch, but it's ultimately TF team's call.

Reply:
Hi Jack, Artem,

Apologies for my delayed response - I was doing some personal traveling last week.

I'd love to discuss more both your goals and how Bazel's multiplatform changes could help them. We're at a weird state now where lots of new possibilities are opening up (like being able to really support per-target crosstools) but the public APIs are still all coming together. So it's not as simple as "just follow this pattern in the Bazel documentation" but that doesn't mean there aren't options for you.

Since I'm just coming in late into this conversation, I suggest we all get on the same basic page of what's desired. Then we can clarify what features can address your goals, and how well.

Would that work?